codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/awssdk/shapevisitor/AwsSdkToDafnyShapeVisitor.java

```diff
@@ -548,15 +550,92 @@ public String doubleShape(DoubleShape shape) {
 
   @Override
   public String unionShape(UnionShape shape) {
+    final ServiceShape serviceShape = context
+        .model()
+        .expectShape(context.settings().getService(context.model()).toShapeId())
+        .asServiceShape().get();
+    final String internalDafnyType = DafnyNameResolver.getDafnyType(
+      shape,
+      context.symbolProvider().toSymbol(shape)
+    );
+    writer.addImportFromModule("github.com/dafny-lang/DafnyRuntimeGo", "dafny");
+    String returnType = DafnyNameResolver.getDafnyType(
+        shape,
+        context.symbolProvider().toSymbol(shape)
+      );
+    String someWrapIfRequired = "%s(%s)";
+    if (this.isOptional) {
+      returnType = "Wrappers.Option";
+      someWrapIfRequired =
+        "Wrappers.Companion_Option_.Create_Some_(%s(%s))";
+    }
+    final String functionInit =
+      """
+      func() %s {
+          switch %s.(type) {""".formatted(returnType, dataSource);
+    StringBuilder eachMemberInUnion = new StringBuilder();
+    for (final var member : shape.getAllMembers().values()) {
+      final String memberName = context.symbolProvider().toMemberName(member);
+      final Shape targetShape = context.model().expectShape(member.getTarget());
+      final String baseType = DafnyNameResolver.getDafnyType(
+        targetShape,
+        context.symbolProvider().toSymbol(targetShape)
+      );
+      String dataSourceInput = dataSource
+                        .concat(".(*")
+                        .concat(SmithyNameResolver.smithyTypesNamespaceAws(serviceShape.expectTrait(ServiceTrait.class), true))
+                        .concat(DOT)
+                        .concat(context.symbolProvider().toMemberName(member))
+                        .concat(").Value");
+      eachMemberInUnion.append(
+        """
+        case *%s.%s:
+            var companion = %s
+            var inputToConversion = %s
+            return %s
+            """.formatted(
+            SmithyNameResolver.smithyTypesNamespaceAws(serviceShape.expectTrait(ServiceTrait.class), true),
+            context.symbolProvider().toMemberName(member),
+            internalDafnyType.replace(
+              shape.getId().getName(),
+              "CompanionStruct_" + shape.getId().getName() + "_{}"
```

Comment:
`.concat()` ?

Reply:
Done here (including a lot of other stuffs) https://github.com/smithy-lang/smithy-dafny/commit/75c8e6a384e4c085e161578cfb682151c7d02e8e